test-utils/src/monerod.rs

```diff
@@ -4,169 +4,139 @@
 //! this to test compatibility with monerod.
 //!
 use std::{
-    collections::HashMap,
+    ffi::OsStr,
+    io::Read,
     net::{IpAddr, Ipv4Addr, SocketAddr},
-    path::PathBuf,
-    process::Stdio,
-    sync::OnceLock,
+    process::{Child, Command, Stdio},
+    str::from_utf8,
+    thread::panicking,
     time::Duration,
 };
 
 use rand::Rng;
-use tokio::{
-    process::{Child, Command},
-    sync::{mpsc, oneshot},
-};
+use tokio::{task::yield_now, time::timeout};
 
 mod download;
 
 const LOCAL_HOST: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
 const MONEROD_VERSION: &str = "v0.18.3.1";
+const MONEROD_STARTUP_TEXT: &str =
+    "The daemon will start synchronizing with the network. This may take a long time to complete.";
 
-#[allow(clippy::type_complexity)]
-static MONEROD_HANDLER_CHANNEL: OnceLock<
-    mpsc::Sender<(MoneroDRequest, oneshot::Sender<(SocketAddr, SocketAddr)>)>,
-> = OnceLock::new();
+const MONEROD_SHUTDOWN_TEXT: &str = "Stopping cryptonote protocol";
 
-/// Spawns monerod and returns the p2p address and rpc address.
-///
-/// When spawning monerod, this module will try to use an already spawned instance to reduce the amount
-/// of instances that need to be spawned.
+/// Spawns monerod and returns [`SpawnedMoneroD`].
 ///
 /// This function will set `regtest` and the P2P/ RPC ports so these can't be included in the flags.
-pub async fn monerod(flags: Vec<String>, mutable: bool) -> (SocketAddr, SocketAddr) {
-    // TODO: sort flags so the same flags in a different order will give the same monerod?
-
-    // We only actually need these channels on first run so this might be wasteful
-    let (tx, rx) = mpsc::channel(3);
-    let mut should_spawn = false;
-
-    let monero_handler_tx = MONEROD_HANDLER_CHANNEL.get_or_init(|| {
-        should_spawn = true;
-        tx
-    });
-
-    if should_spawn {
-        // If this call was the first call to start a monerod instance then start the handler.
-        let manager = MoneroDManager::new().await;
-        tokio::task::spawn(manager.run(rx));
-    }
+pub async fn monerod<T: AsRef<OsStr>>(flags: impl IntoIterator<Item = T>) -> SpawnedMoneroD {
+    let path_to_monerod = download::check_download_monerod().await.unwrap();
+    let mut rng = rand::thread_rng();
+
+    // Use random ports and *hope* we don't get a collision (TODO: just keep an atomic counter and increment?)
+    let rpc_port: u16 = rng.gen_range(1500..u16::MAX);
+    let p2p_port: u16 = rng.gen_range(1500..u16::MAX);
+    let zmq_port: u16 = rng.gen_range(1500..u16::MAX);
```

Comment:
```rust
fn get_available_port(already_taken: &[u16]) -> u16 {
    loop {
        // Using `0` makes the OS return a random available port.
        let port = std::net::TcpListener::bind("127.0.0.1:0")
            .unwrap()
            .local_addr()
            .unwrap()
            .port();

        if !already_taken.contains(&port) {
            return port;
        }
    }
}

let rpc_port = get_available_port(&[]);
let p2p_port = get_available_port(&[rpc_port]);
let zmq_port = get_available_port(&[rpc_port, p2p_port]);
```

Reply:
Ah, but there are multiple `monerod`'s... `already_taken` can be `Arc<Mutex<Vec<u16>>` and updated for all I guess.

Reply:
The OS will see that the ports are taken unless they are started at the same time, I think the chance of this happening is probably low enough to just go with your first idea